Discuss Application-Limited Sending (#1637)
quicwg/base-drafts: draft-ietf-quic-recovery.md

During recovery, the congestion window remains unchanged irrespective of new
losses or increases in the ECN-CE counter.

## Application Limited Sending

If the sender is sufficiently application limited that the congestion window is
not fully utilized, the congestion window should not be increased in slow start
or congestion avoidance.  Senders should consider themselves application limited
if bytes in flight when receiving an ACK frame are more than a max datgram size

> less than the congestion window.

Reviewer: This last sentence is a run-on and unnecessarily detailed. I would simply say something like "if the congestion window is not fully utilized." The detail of kMaxDatagramSize is a rounding thing, which is probably too much detail in this descriptive text.

Author: In that case, I think I should just remove that sentence?

A sender MAY implement alternate mechanisms to update its congestion window
after idle periods, such as those proposed for TCP in {{?RFC7661}}.

## Application Limited Sending

The congestion window should not be increased in slow start or congestion
avoidance when it is not fully utilized.  The congestion window could be
under-utilized due to insufficient application data to send or flow control
limits.

> When the sender is pacing(see {{pacing}}) packets, the sender may be unable

Reviewer 2: ```suggestion
When the sender is pacing (see {{pacing}}) packets, the sender may be unable
```

to use the full congestion window for a period of time after receiving an

> ACK, due to pacing.  In this case, the sender should not consider themselves

Reviewer 2: ```suggestion
acknowledgment.  In this case, the sender should not consider themselves
```

Author: I think I mean ACK, as in the frame?

Reviewer 2: You need more than an ACK, you need one that contains fresh acknowledgments, right?

Author: True, changed to acknowledgements.

> application limited and should allow the congestion window to increase.

Reviewer 2: I'm not really following the logic here.  I think that you need more words.

The reason that the congestion window is limited in this fashion is that without being exercised, there is no assurance that this much data can be sent.  If there is a sudden increase in demand from the application such that the inflated limit is now used, that results in using an untested limit, which might result in severe congestion.

Author: SG, maybe @janaiyengar or @nibanks will have a suggestion?

Reviewer 3: Maybe it's just a case of too much detail. I think folks should understand that just because a pacing interval has currently been used up, that's not considered application limited? So, maybe, just don't have this at all?

A sender MAY implement alternate mechanisms to update its congestion window
after idle periods, such as those proposed for TCP in {{?RFC7661}}.

## Application Limited Sending

The congestion window should not be increased in slow start or congestion
avoidance when it is not fully utilized.  The congestion window could be

> under-utilized due to insufficient application data to send or flow control

Reviewer: ```suggestion
under-utilized due to insufficient application data or flow control credit.
```

> limits.

Reviewer: ```suggestion
```

When the sender is pacing (see {{pacing}}) packets, the sender may be unable
to use the full congestion window for a period of time after receiving
acknowledgements, due to pacing.  In this case, the sender should not consider
themselves application limited and should allow the congestion window to

> increase.

Reviewer: Suggested rephrasing:  "A sender that uses pacing (see {{pacing}}) might delay sending of packets and might not fully utilize the congestion window due to this delay.  A sender should not consider itself application limited if it might have utilized the entire congestion window without pacing delay."

"When the sender paces packets (see {{pacing}}), it might not send 
to use the full congestion window for a period of time after receiving
acknowledgements, due to pacing.  In this case, the sender should not consider
themselves application limited and should allow the congestion window to
increase."

Author: Accepted with minor tweaks.

~~~
   InRecovery(packet_number):
     return packet_number <= end_of_recovery

>    

Reviewer 4: Build failure is trailing whitespace on this line.

   IsAppLimited():

>      bytes_in_flight >= congestion_window - kMaximumDatagramSize

Reviewer 4: Isn't this backward?  Bytes in flight can't be greater than the congestion window, but if the bytes in flight are within one datagram of the window, you're clearly not limited.

Reviewer 3: I agree this is backward. The `>=` should just be `<` or `<=`. But as far as tracking `IsAppLimited()` or whatever you want to call it, I would just suggest a modification to the code:

Instead of using `bytes_in_flight` here and the helper function `IsAppLimited`, introduce a new variable `window_full` that is only set in `OnPacketSent`. At the end of each call to that function, if `bytes_in_flight` is close enough to `congestion_window` set it to true, otherwise false. Then below, you use that variable to decide if the cwnd should be increased..

Reviewer 2: That seems better, yes.  When you say "close enough", I assume that you mean "so close that you couldn't send anything meaningful", something like headerlen+sizeof(cid)+16+sizeof(STREAM frame overhead), so maybe sizeof(cid)+32 ?

Reviewer 3: I was specifically not defining "close enough". I am fine with a single datagram size or something smaller.

~~~
   InRecovery(packet_number):
     return packet_number <= end_of_recovery
   
   IsAppLimited():
     bytes_in_flight >= congestion_window - kMaximumDatagramSize

   OnPacketAckedCC(acked_packet):
     // Remove from bytes_in_flight.
     bytes_in_flight -= acked_packet.bytes
     if (InRecovery(acked_packet.packet_number)):
       // Do not increase congestion window in recovery period.
       return
     if (IsAppLimited())
       // Do not increase congestion_window if application limited.

>        return

Reviewer 3: This gets into implementation specifics a bit, but I don't think the pseudocode can be as trivial as this.

Naively, if this function is called for two different packets in a row, without a corresponding send operation being performed between them, then only the first `OnPacketAckedCC` ever actually increases `congestion_window` because the previous call just decremented `bytes_in_flight` and incremented `congestion_window` guaranteeing that `IsAppLimited()` will always return false for future packets.

In a slightly more intelligent interpretation of this, where you actually preform this logic for all the bytes acknowledged in a whole ACK frame in a single call, you can still have the same problem if you end up processing two ACK frames back to back before you end up doing a send. I have some ideas on how to solve that (at least for my implementation) but I was wondering if this should be mentioned (if not pseudocoded) here.

Author: Agreed, getting this right is complex.  I am going to not define the IsAppLimited() method, but document what it's intent is.
